consumer/consumer.go

```diff
@@ -3,177 +3,193 @@ package consumer
-	// a message is successfully handled.  The Report function will
-	// receive a copy of the message that was handled, along with
-	// a map[string]string containing metrics about the handling of the
-	// message.  Currently we pass the following metrics: "attempts" (the
-	// number of attempts it took before the message was handled);
-	// "msgOffset" (the marked offset for the message on the topic); and
-	// "remainingOffset" (the difference between the high water mark and
-	// the current offset).
-	Metrics MetricsReporter
+	config Config
+
+	// mu guards the fields below.
+	mu sync.Mutex
 
-	// Logger output defaults to ioutil.Discard.
-	// If you wish to provide a different value for the Logger, you must do this prior to calling Consumer.Serve.
-	Logger *log.Logger
+	// quit is closed when the consumer is closed.
+	quit chan struct{}
 
-	consumer sarama.ConsumerGroup
-	config   *Config
-	handlers *collection
-	quit     chan struct{}
+	// consumerGroup holds the consumerGroup started by Serve.
+	consumerGroup sarama.ConsumerGroup
+	// handlers holds the set of handlers.
+	handlers handlers
 }
 
-// New returns a ready to use Consumer configured with sane defaults.
+// New returns a new consumer with the given configuration.
+// Use the Handle method to register handlers for all subscribed
+// topics, then call Serve to begin consuming messages.
+//
+// The Consumer must be closed with the Close method after use.
 func New(cfg Config) (*Consumer, error) {
 	err := cfg.Validate()
 	if err != nil {
 		return nil, err
 	}
-
 	return &Consumer{
-		Logger: log.New(ioutil.Discard, "[Felice] ", log.LstdFlags),
-
-		config:   &cfg,
-		handlers: new(collection),
-		quit:     make(chan struct{}),
+		config: cfg,
+		quit:   make(chan struct{}),
 	}, nil
 }
 
 // Handle registers the handler for the given topic.
-// Handle must be called before Serve.
+// It will panic if it is called after Serve or the topic
+// is empty.
 func (c *Consumer) Handle(topic string, converter MessageConverter, h Handler) {
-	c.handlers.Set(topic, HandlerConfig{
-		Handler:   h,
-		Converter: converter,
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.consumerGroup != nil {
+		panic("Consumer.Handle called after Consumer.Serve")
+	}
+	c.handlers.add(&topicHandler{
+		topic:     topic,
+		handler:   h,
+		converter: converter,
 	})
-
-	c.Logger.Printf("Registered handler. topic=%q\n", topic)
 }
 
-// Serve runs the consumer and listens for new messages on the given topics.
-// Serve will block until it is instructed to stop, which you can achieve by calling Consumer.Stop.
-// When Serve terminates it will return an Error or nil to indicate that it exited without error.
-func (c *Consumer) Serve() error {
-	var err error
+// Serve runs the consumer and listens for new messages on the given
+// topics. Serve will block until it is instructed to stop, which you
+// can achieve by calling Close. It returns an error only if Close
+// was called and the consumer shut down cleanly.
```

Comment:
This second sentence here is somewhat surprising.

Reply:
You mean you'd expect "Close" to be named "Stop" ?
You may be right in that case, but I generally think that `Close` is the name we use in Go for closing resources, and this is really just another resource to be cleaned up.